stl/inc/xutility

```diff
@@ -3063,8 +3063,477 @@ namespace ranges {
     template <class _Rng>
     concept common_range = range<_Rng>
         && same_as<iterator_t<_Rng>, sentinel_t<_Rng>>;
+
+    // CLASS TEMPLATE ranges::view_interface
+    template <class _Ty>
+    concept _Can_empty = requires(_Ty __t) { _RANGES empty(__t); };
+
+    template <class _Derived>
+        requires is_class_v<_Derived> && same_as<_Derived, remove_cv_t<_Derived>>
+    class view_interface : public view_base {
+    private:
+        _NODISCARD constexpr _Derived& _Cast() noexcept {
+            static_assert(derived_from<_Derived, view_interface>,
+                "view_interface's template argument T must derive from view_interface<T> (N4849 [view.interface]/2).");
+            static_assert(view<_Derived>,
+                "view_interface's template argument must model the view concept (N4849 [view.interface]/2).");
+            return static_cast<_Derived&>(*this);
+        }
+        _NODISCARD constexpr const _Derived& _Cast() const noexcept {
+            static_assert(derived_from<_Derived, view_interface>,
+                "view_interface's template argument T must derive from view_interface<T> (N4849 [view.interface]/2).");
+            static_assert(view<_Derived>,
+                "view_interface's template argument must model the view concept (N4849 [view.interface]/2).");
+            return static_cast<const _Derived&>(*this);
+        }
+
+    public:
+#ifdef __clang__ // TRANSITION, LLVM-44833
+        template <class _Dx = _Derived>
+        _NODISCARD constexpr bool empty() requires forward_range<_Dx>

```

Comment:
I appologize if this is obvious but I had a hard time figuring this out so I would like to recapitulate:

We need the `_Can_empty` concept because for the `operator()` we need `empty` and we need to ensure that any class `_Derived` that is a `view` but not a `forwarding_range` yet implements `empty` can be casted.

What I am wondering is a class `_Derived` that *is* a `forwarding_range` *and* implements `empty`. We are safe in the `operator()` due to the explicit cast to `_Derived` as far as I understand the CPO. However, would it make sense to additionally constrain these subfunctions here with `!_Can_empty<_Derived>` so that we do not needlessly shadow it?

Reply:
If the derived class doesn't define `empty()` (and is forward or stronger), then you `_Can_empty` the derived class anyway through this very member function.

Reply:
As Tim is hinting, you're leaping off the edge of constraint recursion:
```c++
template <class D>
struct Base {
  void f() requires (!requires(D d) { d.f(); }) {
    // ...
  }
};

struct Derived : Base<Derived> {};
Derived d;
```
`d.f()` is a valid expression only if `d.f()` is not a valid expression.